# BUDS Report 10: Group


### Table of Contents

1. <a href='#section 1'>Ice Cream Flavors </a>
    
2. <a href='#section 2'> NBA Salaries </a>

In [1]:
# run this cell
from datascience import *
import numpy as np
import math
import matplotlib.pyplot as plt
plt.style.use("fivethirtyeight")
%matplotlib inline

## 1. Ice Cream Flavors <a id='section 1'></a>

Data scientists often need to classify individuals into groups according to shared features and may like to identify some characteristics of the groups. Let's start with a table of ice cream flavors.

|Flavor|Price|
|-|-|
|strawberry|3.55|
|chocolate|4.75|
|chocolate|6.55|
|strawberry|5.25|
|chocolate|5.25|

<div class="alert alert-warning">
    <b>PRACTICE:</b> Create the table depicted above and assign it to a variable called <code>cones</code>. Feel free to reference Report 06, which covers "Creating Tables". </div>

In [ ]:
cones = ...
cones

In [2]:
# SOLUTION
cones = Table().with_columns(
    'Flavor', make_array('strawberry', 'chocolate', 'chocolate', 'strawberry', 'chocolate'),
    'Price', make_array(3.55, 4.75, 6.55, 5.25, 5.25)
)
cones

Flavor     | Price
strawberry | 3.55
chocolate  | 4.75
chocolate  | 6.55
strawberry | 5.25
chocolate  | 5.25

What if we wanted to count the total _number_ of each flavor?

The group method with a single argument counts the number of rows for each category in a column. The result contains one row per unique value in the grouped column.

The format will generally follow `tbl.group(col_to_group_by)`.

<div class="alert alert-warning">
    <b>PRACTICE:</b> Try calling the <code>group</code> method to see what this looks like. </div>

In [ ]:
...

In [3]:
# SOLUTION
cones.group('Flavor')

Flavor     | count
chocolate  | 3
strawberry | 2

What if we instead wanted the *total price* of the cones of each different flavor? That's where the second argument of group comes in.

The optional second argument names the function that will be used to aggregate values in other columns. For instance, sum will sum up the prices in all rows that match each category. This result also contains one row per unique value in the grouped column, but it has the same number of columns as the original table.

This call will follow a similar format as above: `tbl.group(col_to_gorup_by, agg_func)`.

<div class="alert alert-warning">
    <b>PRACTICE:</b> Again, try calling the <code>group</code> method to see what this looks like. Make sure to use the second argument.
</div>

In [ ]:
...

In [4]:
# SOLUTION
cones.group('Flavor', sum)

Flavor     | Price sum
chocolate  | 16.55
strawberry | 8.8

To create this new table, group has calculated the sum of the "Price" entries in all the rows corresponding to each distinct flavor. The prices in the three chocolate rows add up to  16.55. The prices in the two strawberry rows have a total of  8.80.

You can replace sum by **any other functions that work on arrays**.

<div class="alert alert-warning">
    <b>PRACTICE:</b> Use grouping to find the highest price in each flavor category.
    </div>
    
_Hint:_ Think of a function that will return the highest price.

In [ ]:
...

In [5]:
# SOLUTION
cones.group('Flavor', max)

Flavor     | Price max
chocolate  | 6.55
strawberry | 5.25

<div class="alert alert-warning">
    <b>PRACTICE:</b> Now use grouping to find the <i>lowest</i> price in each flavor category. Assign it to a variable called <code>cheaper</code>.</div>

In [ ]:
...

In [6]:
# SOLUTION
cheaper = cones.group('Flavor', min)
cheaper

Flavor     | Price min
chocolate  | 4.75
strawberry | 3.55

<div class="alert alert-warning">
    <b>PRACTICE:</b> Write code that evaluates to the flavor with the highest lowest price using the table from above. What data type should be outputted?</div>

In [ ]:
...

In [7]:
# SOLUTION
sorted_cones = cheaper.sort('Price min', descending=True)
sorted_cones.column('Flavor').item(0)

'chocolate'

## 2. NBA Salaries <a id='section 2'></a>

Now that you've played with a toy dataset, let's work with a real dataset. We can look at some data about NBA players where salaries are measured in millions of dollars.

In [8]:
nba = Table.read_table('nba_salaries.csv')
nba

PLAYER           | POSITION | TEAM          | 2015-2016 SALARY
Paul Millsap     | PF       | Atlanta Hawks | 18.6717
Al Horford       | C        | Atlanta Hawks | 12
Tiago Splitter   | C        | Atlanta Hawks | 9.75625
Jeff Teague      | PG       | Atlanta Hawks | 8
Kyle Korver      | SG       | Atlanta Hawks | 5.74648
Thabo Sefolosha  | SF       | Atlanta Hawks | 4
Mike Scott       | PF       | Atlanta Hawks | 3.33333
Kent Bazemore    | SF       | Atlanta Hawks | 2
Dennis Schroder  | PG       | Atlanta Hawks | 1.7634
Tim Hardaway Jr. | SG       | Atlanta Hawks | 1.30452
... (407 rows omitted)

<div class="alert alert-warning">
    <b>PRACTICE:</b> Relabel the salaries column to something shorter, so that it's easier to work with. </div>

In [ ]:
...

In [9]:
# SOLUTION
nba = nba.relabeled('2015-2016 SALARY', 'SALARY')
nba

PLAYER           | POSITION | TEAM          | SALARY
Paul Millsap     | PF       | Atlanta Hawks | 18.6717
Al Horford       | C        | Atlanta Hawks | 12
Tiago Splitter   | C        | Atlanta Hawks | 9.75625
Jeff Teague      | PG       | Atlanta Hawks | 8
Kyle Korver      | SG       | Atlanta Hawks | 5.74648
Thabo Sefolosha  | SF       | Atlanta Hawks | 4
Mike Scott       | PF       | Atlanta Hawks | 3.33333
Kent Bazemore    | SF       | Atlanta Hawks | 2
Dennis Schroder  | PG       | Atlanta Hawks | 1.7634
Tim Hardaway Jr. | SG       | Atlanta Hawks | 1.30452
... (407 rows omitted)

Before conducting exploratory data analysis on this dataset, answer a few questions about the data.

1. What are the different positions in an NBA team, and what are their abbreviations in this table?

2. How many teams are there in the NBA that are represented in this table?

_Written Answer:_

The aggregate function that we pass as our second argument might not work for all of our columns in the dataset. For example, we can't take the average of a column with strings.

Let's see what happens when we tell the computer to take the average of a column of strings. 

<div class="alert alert-warning">
    <b>PRACTICE:</b> Use average as the aggregate function in the <code>group</code> method. Be sure that the aggregate function will be called on a column made of strings. What happens? Does the cell error? Does something show up?</div>

In [ ]:
...

In [10]:
# SOLUTION
nba.group('TEAM', np.average)

TEAM                  | PLAYER average | POSITION average | SALARY average
Atlanta Hawks         |                |                  | 4.96951
Boston Celtics        |                |                  | 3.35237
Brooklyn Nets         |                |                  | 4.40823
Charlotte Hornets     |                |                  | 4.67236
Chicago Bulls         |                |                  | 6.56841
Cleveland Cavaliers   |                |                  | 10.2312
Dallas Mavericks      |                |                  | 5.97841
Denver Nuggets        |                |                  | 4.45924
Detroit Pistons       |                |                  | 4.22118
Golden State Warriors |                |                  | 6.72037
... (20 rows omitted)

To make coding in the future easier, take only the necessary columns from our table before performing table manipulations like this. You can do so in multiple lines of code. Try this in the next practice question.

<div class="alert alert-warning">
    <b>PRACTICE:</b> How much money did each team pay for its players' salaries? </div>

In [ ]:
...

In [11]:
# SOLUTIONS
teams_and_money = nba.select('TEAM', 'SALARY')
teams_and_money.group('TEAM', sum)

TEAM                  | SALARY sum
Atlanta Hawks         | 69.5731
Boston Celtics        | 50.2855
Brooklyn Nets         | 57.307
Charlotte Hornets     | 84.1024
Chicago Bulls         | 78.8209
Cleveland Cavaliers   | 102.312
Dallas Mavericks      | 65.7626
Denver Nuggets        | 62.4294
Detroit Pistons       | 42.2118
Golden State Warriors | 94.0851
... (20 rows omitted)

Taking that extra step makes our tables a lot cleaner.

Next, let's look at a characteristic that `group` has.

<div class="alert alert-warning">
    <b>PRACTICE:</b> Create a table that shows how many NBA players were in each of the five positions. Take note of the order of the values under the column we grouped by. What order does it seem to follow? Is it based on the original table's order?</div>

In [ ]:
...

In [12]:
# SOLUTIONS
nba.group('POSITION')

POSITION | count
C        | 69
PF       | 85
PG       | 85
SF       | 82
SG       | 96

_Written Answer:_

<div class="alert alert-warning">
    <b>PRACTICE:</b> What was the average salary of the players at each of the five positions? Why do you think centers get paid the most?</div>

In [ ]:
...

In [13]:
# SOLUTIONS
positions_and_money = nba.select('POSITION', 'SALARY')
positions_and_money.group('POSITION', np.average)

POSITION | SALARY average
C        | 6.08291
PF       | 4.95134
PG       | 5.16549
SF       | 5.53267
SG       | 3.9882

_Written Answer:_

<div class="alert alert-warning">
    <b>PRACTICE:</b> Find the highest player salary for each team, then sort the teams in order of highest player salaries to lowest. What types of teams pay the highest individual salaries?</div>
    
*Hint:* Think about which columns you need, what you would group by, and finally what you would sort by.

In [ ]:
...

In [15]:
# SOLUTIONS
teams_and_money = nba.select('TEAM', 'SALARY')
highest_salaries = teams_and_money.group('TEAM', max)
highest_salaries.sort('SALARY max', descending=True)

TEAM                  | SALARY max
Los Angeles Lakers    | 25
Brooklyn Nets         | 24.8949
Cleveland Cavaliers   | 22.9705
New York Knicks       | 22.875
Houston Rockets       | 22.3594
Miami Heat            | 22.1927
Los Angeles Clippers  | 21.4687
Oklahoma City Thunder | 20.1586
Chicago Bulls         | 20.0931
San Antonio Spurs     | 19.689
... (20 rows omitted)

_Written Answer:_

In order to get a better understanding of how group operates under the hood, let's re-create the grouping process with code that we already knew how to do.

<div class="alert alert-warning">
    <b>PRACTICE:</b> Using the <code>group</code> method, create a table that looks like the one depicted below. Make sure that it doesn't have any additional rows.
    
Then, try to think of a way that you would have manually found this information before. For example, ask yourself how you would have found out the number of Atlanta Hawks players. Referring to Report 09 might help. Once you've acquired all of the necessary information, recreate this table.</div>
    
TEAM|count
-|-
Atlanta Hawks|14
Boston Celtics|15

In [ ]:
# method 1:
...

In [16]:
# SOLUTION
nba.group('TEAM').take(0, 1)

TEAM           | count
Atlanta Hawks  | 14
Boston Celtics | 15

In [ ]:
# method 2:
...

In [17]:
# SOLUTION
num_atl = nba.where('TEAM', 'Atlanta Hawks').num_rows
num_bost = nba.where('TEAM', 'Boston Celtics').num_rows

small_group = Table().with_columns('TEAM', make_array('Atlanta Hawks', 'Boston Celtics'),
                                   'count', make_array(num_atl, num_bost))
small_group

TEAM           | count
Atlanta Hawks  | 14
Boston Celtics | 15

`group` expedites that process, making it much more wieldy to do that process for columns with many unique items.

However, its usefulness extends past that! The group method has even more capability.

<div class="alert alert-warning">
    <b>PRACTICE:</b> In the following cell, create a table that show the average salary for each position on each team. Try thinking of how this might be done without code or think about what the resulting table might look like.

This exercise will show another nice property of grouping.</div>

In [ ]:
...

In [18]:
# SOLUTION
team_pos_money = nba.select('TEAM', 'POSITION', 'SALARY')
team_salary_pos = team_pos_money.group(make_array('TEAM', 'POSITION'), np.average)
team_salary_pos

/opt/conda/lib/python3.9/site-packages/datascience/tables.py:920: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  values = np.array(tuple(values))


TEAM           | POSITION | SALARY average
Atlanta Hawks  | C        | 7.58542
Atlanta Hawks  | PF       | 11.0025
Atlanta Hawks  | PG       | 4.8817
Atlanta Hawks  | SF       | 3
Atlanta Hawks  | SG       | 1.80969
Boston Celtics | C        | 2.45046
Boston Celtics | PF       | 3.08548
Boston Celtics | PG       | 4.97465
Boston Celtics | SF       | 4.41716
Boston Celtics | SG       | 2.00755
... (137 rows omitted)

Let's try to find some more information on each team's payments to the different positions.

<div class="alert alert-warning">
    <b>PRACTICE:</b> For each team, find the position it pays the most on average.
    
This will take <i>multiple</i> steps, so think about what this process might look like with the Atlanta Hawks in the table you just made above. What position do the Hawks pay the most on average? How did you get that value? </div>

In [ ]:
...

In [19]:
# SOLUTION
team_salary_pos.sort('SALARY average', descending=True).sort('TEAM', distinct=True)

TEAM                  | POSITION | SALARY average
Atlanta Hawks         | PF       | 11.0025
Boston Celtics        | PG       | 4.97465
Brooklyn Nets         | SF       | 13.0403
Charlotte Hornets     | C        | 6.77224
Chicago Bulls         | PG       | 11.1715
Cleveland Cavaliers   | SF       | 22.9705
Dallas Mavericks      | SF       | 15.3615
Denver Nuggets        | SF       | 7.19577
Detroit Pistons       | PG       | 13.913
Golden State Warriors | SG       | 9.0005
... (20 rows omitted)

### Downloading as PDF

Download this notebook as a pdf by clicking <b><code>File > Download as > PDF via LaTeX (.pdf)</code></b>. Turn in the PDF into bCourses under the corresponding assignment.

## Bibliography

- Data 8 Textbook — Classifying by One Variable: https://www.inferentialthinking.com/chapters/08/2/Classifying_by_One_Variable.html
---
Notebook developed by: 

Data Science Modules: http://data.berkeley.edu/education/modules